fpdf/fpdf.py

```diff
@@ -3155,6 +3172,7 @@ def will_page_break(self, height):
         return (
             self.y + height > self.page_break_trigger
             and not self.in_footer
+            and self.page_body_has_content
```

Comment:
I'm not a big fan of introducing a new attribute on the `FPDF` class...
I already has too many properties.
I have tested the following code and I think it should behave similarly:
```python
        return (
            self.y > self.t_margin  # ensure that there is already some content on the page
            and self.y + height > self.page_break_trigger
            and not self.in_footer
            and self.accept_page_break
        )
```

Reply:
> I'm not a big fan of introducing a new attribute on the `FPDF` class... It already has too many properties. I have tested the following code and I think it should behave similarly:
> 
> ```python
>         return (
>             self.y > self.t_margin  # ensure that there is already some content on the page
>             and self.y + height > self.page_break_trigger
>             and not self.in_footer
>             and self.accept_page_break
>         )
> ```

The problem with this is that it wouldn't work when y is set before inserting the element (which I for example often do, but that might just be me). I couldn't see a way around adding a new attribute that would cover all use cases

Reply:
> The problem with this is that it wouldn't work when y is set before inserting the element (which I for example often do, but that might just be me). I couldn't see a way around adding a new attribute that would cover all use cases

OK I see.

Then maybe the heuristic could be this:
```python
        # The magic number 11 just below is len(bytearray(b'2 J\n0.57 w\n')),
        # the value of self.pages[self.page]["content"] just after _beginpage() has been called:
        page_has_content = self.y > self.t_margin and self.page > 0 and len(self.pages[self.page]["content"]) > 11
        return (
            page_has_content
            and self.y + height > self.page_break_trigger
            and not self.in_footer
            and self.accept_page_break
        )
```

What do you think @semaeostomea & @gmischler?

Reply:
Independently of whether there's already other content on a page, `self.y <= self.t_margin` might still be a valid test for not triggering an auto page break. After all, if an element won't fit on the page under those circumstances, then it won't fit any better on the next one either.

The auto-page-break feature was designed for flowing text content, and I don't expect it to ever trigger prematurely for any reasonably sized text, or even for a reasonably sized image.
Now if - on an empty page - you explicitly set `self.y` to a position lower than `self.t_margin` before inserting something huge, I'd say it is your own job to check first that it actually fits (or just disable auto breaking to begin with).

The "problem" being "solved" here really can only manifest itself in an already pathological situation, which probably doesn't justify the hassle. I'm not so much worried about an extra instance attribute, but maintaining this one in a reliable manner appears to be a rather messy task. It is already tricky to implement correctly, and will be highly error prone to work with in the context of a subclass or other ways to extend the library. And however we try to do it, it will probably fail when a user starts moving back and forth between pages, which is a perfectly valid thing to do!

Given all that, my vote would be to shelve this idea for the moment. Given the way how fpdf2 is currently structured, making automatic page breaks absolutely fool proof is not a realistic undertaking. And this implementation may cause as many new problems as it solves existing ones.

Reply:
I agree with you @gmischler 

I introduced the `self.y > self.t_margin` check in #535

I'm going to wait for @semaeostomea answer/comments now.
Let's discuss if we should do more about this or just accept it as an edge case that end users should deal with on their own

Reply:
Btw.: Unfortunately, checking for those magic 11 bytes above doesn't reliably work either.  This test will give the wrong result when the user has changed some settings like the font or a color. Many of those changes get immediately written to the output stream for the current page, so that its contents will then be longer even though nothing visible has been added to the page.

Fpdf2 is currently a very open system, which is really a feature, as it gives it a lot of flexibility and makes it straightforward to use. At the same time some automated functions are offered, but they make the system at best "semi automated". It is hard to put guard rails around stuff, because the content streams in a PDF file are just an unstructured sequence of text instructions. There are places where we could probably add more structured levels of abstraction and assert a more tight control over individual settings (for text content some of that has already happened), but its not an easy task. 

So maybe we could say that this proposal comes a bit too early in the evolution of the library. Sorry for that... :wink:

Reply:
> I also intended for oversized images to get cut off on all sides and aligned them to the center.

If you already know that your content won't fit the page in the first place and should be cropped, then the correct way to handle that (short of actually cropping it) is to turn off auto-breaking and create the pages new manually.

Given the current architecture of the library, I don't see a way to make this idea fool-proof and not end up with a maintenance nightmare. There are just too many ways a user can circumvent the automation features (including direct access to `._out()`, even if that may not be specifically encouraged).

Thanks for trying anyway, though!

Reply:
@Lucas-C just to double-check since you neither told me to go ahead nor anyone closed the PR:
should I go ahead with my idea of splitting up the page content or is this idea shelved for the moment and the PR can be closed?

Reply:
> should I go ahead with my idea of splitting up the page content or is this idea shelved for the moment and the PR can be closed?

I honestly would prefer <ins>not</ins>to go this road for now, so I'm going to close this PR.
Thank you very much @semaeostomea for contributing this suggestion!